test/integration/test_server.c

```diff
@@ -0,0 +1,262 @@
+#include "../../include/dqlite.h"
+#include "../lib/fs.h"
+#include "../lib/munit.h"
+#include "../lib/runner.h"
+
+#include <stdio.h>
+#include <unistd.h>
+
+SUITE(server);
+
+#define N_SERVERS 3
+
+struct fixture
+{
+	char *dirs[N_SERVERS];
+	dqlite_server *servers[N_SERVERS];
+};
+
+static void *setup(const MunitParameter params[], void *user_data)
+{
+	(void)params;
+	(void)user_data;
+	struct fixture *f = munit_malloc(sizeof *f);
+	unsigned i;
+	int rv;
+
+	for (i = 0; i < N_SERVERS; i += 1) {
+		f->dirs[i] = test_dir_setup();
+		rv = dqlite_server_create(f->dirs[i], &f->servers[i]);
+		munit_assert_int(rv, ==, 0);
+	}
+
+	return f;
+}
+
+static void teardown(void *data)
+{
+	struct fixture *f = data;
+	unsigned i;
+
+	for (i = 0; i < N_SERVERS; i += 1) {
+		dqlite_server_destroy(f->servers[i]);
+		test_dir_tear_down(f->dirs[i]);
+	}
+	free(f);
+}
+
+#define PREPARE_FILE(i, name, s, ...)                           \
+	do {                                                    \
+		char path[100];                                 \
+		snprintf(path, 100, "%s/%s", f->dirs[i], name); \
+		FILE *fp = fopen(path, "w+");                   \
+		fprintf(fp, s, __VA_ARGS__);                    \
+		fclose(fp);                                     \
+	} while (0)
+
+#define NODE(x) x
+
+#define NODE0_ID "3297041220608546238"
+
+void start_each_server(struct fixture *f)
+{
+	int rv;
+
+	rv = dqlite_server_set_address(f->servers[0], "127.0.0.1:8880");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_bootstrap(f->servers[0]);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_refresh_period(f->servers[0], 1);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_start(f->servers[0]);
+	munit_assert_int(rv, ==, 0);
+
+	rv = dqlite_server_set_address(f->servers[1], "127.0.0.1:8881");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_peer_address(f->servers[1], "127.0.0.1:8880");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_refresh_period(f->servers[0], 1);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_start(f->servers[1]);
+	munit_assert_int(rv, ==, 0);
+
+	rv = dqlite_server_set_address(f->servers[2], "127.0.0.1:8882");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_peer_address(f->servers[2], "127.0.0.1:8880");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_peer_address(f->servers[2], "127.0.0.1:8881");
+	munit_assert_int(rv, ==, 0);
```

Comment:
From the point of view of somebody not familiar with dqlite, this sequence of calls to `dqlite_server_set_peer_address()` seems to be a bit ambiguous.

Usually one would expect that a call like `my_object_set_x(object, some_value)` would set the value of `x` to `some_value` on `object`, and a subsequent call like `my_object_set_x(object, other_value)` would replace the previously set `some_value` with `other_value`.

Also, one would expect that calling `my_object_set_x(object, some_value)` two times in a row is idempotent.

However the current semantics of `dqlite_server_set_peer_address()` is different from the "intuitive" one, and one must read the documentation to understand the details. For example calling `dqlite_server_set_peer_address(server, "127.0.0.1:8880")` two times is not idempotent, and calling it with another address does not replace the previous address. The current semantics is also somehow different from the one of the other `dqlite_server_set_xxx()`, which are idempotent and overwrite previous values.

What about replacing `dqlite_server_set_peer_address()` with `dqlite_server_set_peer_addresses()`? Or even `dqlite_server_set_peers()`, for brevity. E.g.

```C
int dqlite_server_set_peers(dqlite_server *s, const char *addresses[], unsigned n);

const char *peers[] = {"127.0.0.1:8880", "127.0.0.1:8881"};
rv = dqlite_server_set_peers(f->servers[2], peers, 2);
munit_assert_int(rv, ==, 0);
```

This would have the intuitive idempotent and overwriting semantics, like the other `set` functions.

Reply:
Maybe also:

```C
int dqlite_server_set_auto_join(dqlite_server *s, const char *peers[], unsigned n);
```

would be even a bit more intuitive. The function turns on the behavior of automatically joining a cluster which has the given peers.